opendata_module/opmon_postgresql_maintenance/create_users.py

```diff
@@ -28,34 +28,103 @@
 
 import argparse
 import getpass
+import logging
+import secrets
+import string
+from typing import Dict
+
 import psycopg2
 import yaml
-from psycopg2.extensions import ISOLATION_LEVEL_AUTOCOMMIT
-from psycopg2.extensions import QuotedString
-import string
-import secrets
+from psycopg2 import errors as psycopg2_errors
+from psycopg2.extensions import (ISOLATION_LEVEL_AUTOCOMMIT,
+                                 ISOLATION_LEVEL_DEFAULT, QuotedString)
+
+logger = logging.getLogger(__file__)
+logger.setLevel(logging.INFO)
+logger.addHandler(logging.StreamHandler())
 
-full_users = {"anonymizer"}
-read_only_users = {"opendata", "networking"}
+
+full_users = {'anonymizer'}
+read_only_users = {'opendata', 'networking'}
 
 
 def main():
     args = _parse_args()
-    postgres = _connect_postgres(args)
-    passwords = _create_users(args, postgres)
-    _create_database(args, postgres)
-    _grant_privileges(args, postgres)
-    _print_users(passwords)
+    postgres_conn = _connect_postgres(args)
+
+    # it is not allowed to create database in transaction context
+    postgres_conn.set_isolation_level(ISOLATION_LEVEL_AUTOCOMMIT)
+    try:
+        _create_database(args, postgres_conn)
+        logger.info('Database created successfully')
+    except psycopg2_errors.DuplicateDatabase as e:
+        logger.info(f'Nothing to do: {str(e)}')
+
+    if args.remove_existing_users:
+        users_to_remove = [f'{user}_{args.xroad}' for user in full_users | read_only_users]
+        logger.info(f'Will try to remove DB users {users_to_remove}\n')
+        _remove_users(args, postgres_conn)
+
+    # create users and grant permissions in single transaction
+    postgres_conn.set_isolation_level(ISOLATION_LEVEL_DEFAULT)
+    try:
+        passwords = _create_users(args, postgres_conn)
```

Comment:
Wait, why `_create_users` return `passwords`? 🤔

Reply:
Because this script is used by sysadmins and generated passwords are used to login DB users.